packages/design-tokens-schema/src/resolve-refs.ts

```diff
@@ -1,66 +1,82 @@
- * extractRefPath('something-else') // => null
- * ```
+ * A predicate function that checks if a given ref in our schema is valid.
+ * Note that it either throws an error or returns a boolean:
+ * This is because it's used both as a Zod validation as well as a regular predicate.
+ *
+ * @param data A part of the root config in which refs may be present
+ * @param root The root of the config in which tokens are searched
  */
-const extractRefPath = (value: string): string | null => {
-  const match = REF_REGEX.exec(value);
-  return match?.[1] ?? null;
-};
+const isTokenWithRef = (data: unknown, root?: Record<string, unknown>): data is TokenWithRef => {
+  // Check that we're dealing with a token-like object
+  const parsedSource = TokenWithRefSchema.safeParse(data);
+  if (parsedSource.success === false) {
+    return false;
+  }
 
-const processRefs = (
-  config: unknown,
-  root: Record<string, unknown> | undefined,
-  onRefFound: (config: ValueObject, key: string, resolvedRef: unknown, tokenType: unknown, refPath: string) => boolean,
-): boolean => {
-  if (!isValueObject(config)) return true;
+  // Grab the `{path.to.ref} -> path.to.ref` and find it inside root
+  const refPath = parsedSource.data.$value.slice(1, -1);
+  const ref = dlv(root, refPath) || dlv(root, `brand.${refPath}`) || dlv(root, `common.${refPath}`);
 
-  for (const key in config) {
-    const value = config[key];
-    if (key === '$value' && typeof value === 'string') {
-      // extract `ma.color.indigo.5` from `{ma.color.indigo.5}` if possible
-      const refPath = extractRefPath(value);
-      if (refPath) {
-        const tokenType = config['$type']; // 'color' | 'fontFamily' | etc.
+  // Check that we're dealing with a token-like object
+  const parsedRef = ReferencedTokenSchema.safeParse(ref);
+  if (parsedRef.success === false) {
+    throw new Error(`Invalid token reference: expected "${refPath}" to have a "$value" and "$type" property`);
+  }
 
-        // does 'ma.color.indigo.5.$value' path exist in `root`?
-        const resolvedRef = dlv(root, refPath);
-        if (!onRefFound(config, key, resolvedRef, tokenType, refPath)) {
-          return false;
-        }
-      }
-    } else if (!processRefs(value, root, onRefFound)) {
-      return false;
-    }
+  // make sure the $type of the referenced token is the same
+  if (parsedSource.data.$type !== parsedRef.data.$type) {
+    throw new Error(
+      `Invalid token reference: $type "${parsedSource.data['$type']}" of "${JSON.stringify(parsedSource.data)}" does not match the $type on reference {${refPath}}. Types "${parsedSource.data.$type}" and "${parsedRef.data.$type}" do not match.`,
+    );
   }
 
   return true;
 };
 
+const walkTokensWithRef = (
+  config: unknown,
+  root: Record<string, unknown>,
+  callback: (token: TokenWithRef) => void,
+): void => {
+  walkObject<TokenWithRef>(config, (data): data is TokenWithRef => isTokenWithRef(data, root), callback);
+};
+
```

Comment:
Maybe move this one to the `walker.ts` since `walkColors` is also exported from there